File: src/main/java/mods/eln/sim/mna/component/Capacitor.java
package mods.eln.sim.mna.component;

import mods.eln.sim.mna.SubSystem;
import mods.eln.sim.mna.misc.ISubSystemProcessI;
import mods.eln.sim.mna.state.State;

public class Capacitor extends Bipole implements ISubSystemProcessI {

    private double coulombs = 0;
    double coulombsPerStep;

    public Capacitor() {}

    public Capacitor(State aPin, State bPin) {
        connectTo(aPin, bPin);
    }

    @Override
    public double getCurrent() {
        return 0;
    }

    public void setCoulombs(double coulombs) {
        this.coulombs = coulombs;
        dirty();
    }

    @Override
    public void applyToSubsystem(SubSystem s) {
        coulombsPerStep = coulombs / s.getDt();

        s.addToA(aPin, aPin, coulombsPerStep);
        s.addToA(aPin, bPin, -coulombsPerStep);
        s.addToA(bPin, bPin, coulombsPerStep);
        s.addToA(bPin, aPin, -coulombsPerStep);
    }

    @Override
    public void simProcessI(SubSystem s) {
        double add = (s.getXSafe(aPin) - s.getXSafe(bPin)) * coulombsPerStep;
        s.addToI(aPin, add);
        s.addToI(bPin, -add);
    }

    @Override
    public void quitSubSystem() {
        subSystem.removeProcess(this);
        super.quitSubSystem();
    }

    @Override
    public void addToSubsystem(SubSystem s) {
        super.addToSubsystem(s);
        s.addProcess(this);
    }

    /**
     * getEnergy
     * (V^2 * C) / 2 = E
     *
     * @return energy, in joules
     */
    public double getEnergy() {
        double voltage = getVoltage();
        return voltage * voltage * coulombs / 2;
    }

    public double getCoulombs() {
        return coulombs;
    }
}

File: src/main/java/mods/eln/sim/mna/component/Capacitor.kt
package mods.eln.sim.mna.component

import mods.eln.Eln
import mods.eln.debug.DebugType
import mods.eln.sim.mna.SubSystem
import mods.eln.sim.mna.misc.ISubSystemProcessI
import mods.eln.sim.mna.state.State

open class Capacitor : Bipole, ISubSystemProcessI {

    var c = 0.0
        set(c) {
            if (c == java.lang.Double.NaN)
                Eln.dp.println(DebugType.MNA, "component.Capacitor setC(double c) - c was NaN!")
            field = c
            dirty()
        }
    internal var cdt: Double = 0.toDouble()

    override fun getCurrent(): Double {
        return 0.0
    }

    fun getE() = getVoltage() * getVoltage() * this.c / 2

    constructor() {}

    constructor(name: String) {
        this.name = name
    }

    constructor(aPin: State, bPin: State) {
        this.name = "Capacitor"
        connectTo(aPin, bPin)
    }

    constructor(name: String, aPin: State, bPin: State) {
        this.name = name
        connectTo(aPin, bPin)
    }


    override fun applyTo(s: SubSystem) {
        cdt = this.c / s.dt

        s.addToA(aPin, aPin, cdt)
        s.addToA(aPin, bPin, -cdt)
        s.addToA(bPin, bPin, cdt)
        s.addToA(bPin, aPin, -cdt)
    }

    override fun simProcessI(s: SubSystem) {
        val add = (s.getXSafe(aPin) - s.getXSafe(bPin)) * cdt
        s.addToI(aPin, add)
        s.addToI(bPin, -add)
    }

    override fun quitSubSystem() {
        getSubSystem()!!.removeProcess(this)
        super.quitSubSystem()
    }

    override fun addedTo(s: SubSystem) {
        super.addedTo(s)
        s.addProcess(this)
    }
}
